File: grokking_algorithms/chapter_6/breadth-first.go
package main

import (
	"fmt"
)

var graph = make(map[string][]string)

func init() {
	graph["you"] = []string{"alice", "bob", "claire"}
	graph["bob"] = []string{"anuj", "peggy"}
	graph["alice"] = []string{"peggy"}
	graph["claire"] = []string{"thom", "jonny"}
	graph["anuj"] = []string{}
	graph["peggy"] = []string{}
	graph["thom"] = []string{}
	graph["jonny"] = []string{}
}

func person_is_seller(name string) bool {
	return name[len(name)-1:] == "m"
}

func search(name string) bool {
	var search_queue = []string{}
	var searched = make(map[string]int)
	search_queue = append(search_queue, graph[name]...)
	// fmt.Printf("%v\n", search_queue)

	for len(search_queue) > 0 {
		person := search_queue[0]       // obtains the first element
		search_queue = search_queue[1:] // update
		// fmt.Println(person, search_queue)
		if _, ok := searched[person]; !ok {
			// fmt.Println("->", ok)
			if person_is_seller(person) {
				fmt.Println(person, " is a mango seller")
				return true
			} else {
				search_queue = append(search_queue, graph[person]...)
				searched[person] = 1
			}
		}
	}
	return false
}

func main() {
	_ = search("you")
}

File: grokking_algorithms/chapter_6/breadth_first.py
from collections import deque

graph = {}
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

def person_is_seller(name):
    return name[-1] == 'm'

def search(name):
    search_queue = deque()
    search_queue += graph[name]
    searched = set() # <- This set is how you keep track of which people you’ve searched before.
    while search_queue:
        person = search_queue.popleft()
        if not person in searched: # <- Only search this person if you haven’t already searched them.
            if person_is_seller(person):
                print(person + " is a mango seller!")
                return True
            else:
                search_queue += graph[person]
                searched.add(person) # <- Marks this person as searched
    return False

search("you")